Fix mapreduce_impl() for 1-element range

alyst opened the pull request:
IIUC, 1-element case is handled by `Base._mapreduce()`, so Base never calls `mapreduce_impl()` for 1-element ranges, but it gets exposed by  `NullableArrays`:
```julia
mean(NullableArray([1]), skipnull=true)
```
```
ERROR: BoundsError: attempt to access 1-element Array{Int64,1} at index [2]
```

Also affects 0.5.

alyst committed:
fix mapreduce_impl() for 1-element range

alyst commented:
I think the linux test failures are unrelated.

tkelman commented:
Can this be tested with a Base (or test-locally-defined) type?

alyst commented:
Late for 0.5.1, but finally managed to add `Base.mapreduce_impl()` tests.

alyst commented:
@tkelman Is it ok to test `mapreduce_impl()` like this?

tkelman commented:
looks reasonable, I think - not too familiar with this code. could also test for type equality with `===`

alyst committed:
mapreduce_impl() type stability tests

alyst commented:
I've added explicit type stability tests.

cc @nalimilan

alyst commented:
@tkelman do you think it is ready to merge now?

tkelman committed:
wrap long lines

tkelman commented on base/reduce.jl:
```diff
@@ -168,10 +168,12 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
 function mapreduce_impl(f, op, A::AbstractArray, ifirst::Integer, ilast::Integer, blksize::Int=pairwise_blocksize(f, op))
-    if ifirst + blksize > ilast
+    if ifirst == ilast
+        @inbounds return r_promote(op, f(A[ifirst]))
```
does this call `op` on the result of `f` or not?

alyst replied:
AFAIU  `r_promote()` method defined in Base does not call `op`. But the line in question matches what `_mapreduce()` does in `n==1` case.

alyst replied:
Actually, there's some code duplication between `_mapreduce()` and `mapreduce_impl()`. To fix that, `_mapreduce()` may call `mapreduce_impl()` for all but empty ranges.

alyst committed:
mapreduce_impl(): add @inbounds

ifirst:ilast should be a valid A indices range (checked by the caller)

alyst commented:
Weird. I've given a try at removing the duplicate code from `_mapreduce()` with an innocent looking commit 2541655, but now the PR fails at bootstrap type with a strange error:
```
92.680954 REPL.jl
error during bootstrap:
LoadError(at "sysimg.jl" line 315: LoadError(at "REPL.jl" line 415: Base.MethodError(f=Base.#isless(), args=(-1, nothing), world=0x0000000000002ef7)))
rec_backtrace at /home/travis/build/JuliaLang/julia/src/stackwalk.c:84
record_backtrace at /home/travis/build/JuliaLang/julia/src/task.c:239
jl_throw at /home/travis/build/JuliaLang/julia/src/task.c:565
eval at /home/travis/build/JuliaLang/julia/src/interpreter.c:501
jl_interpret_toplevel_expr at /home/travis/build/JuliaLang/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:477
jl_eval_module_expr at /home/travis/build/JuliaLang/julia/src/toplevel.c:203
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:480
jl_parse_eval_all at /home/travis/build/JuliaLang/julia/src/ast.c:873
jl_load at /home/travis/build/JuliaLang/julia/src/toplevel.c:616
_include at ./sysimg.jl:241
unknown function (ip: 0x2b0ae9f35213)
jl_call_fptr_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/travis/build/JuliaLang/julia/src/gf.c:2225
include at ./sysimg.jl:12
unknown function (ip: 0x2b0aea281c93)
jl_call_fptr_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/travis/build/JuliaLang/julia/src/gf.c:2225
do_call at /home/travis/build/JuliaLang/julia/src/interpreter.c:75
eval at /home/travis/build/JuliaLang/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/travis/build/JuliaLang/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:577
jl_eval_module_expr at /home/travis/build/JuliaLang/julia/src/toplevel.c:203
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:480
jl_parse_eval_all at /home/travis/build/JuliaLang/julia/src/ast.c:873
jl_load at /home/travis/build/JuliaLang/julia/src/toplevel.c:616
exec_program at /home/travis/build/JuliaLang/julia/ui/repl.c:49
true_main at /home/travis/build/JuliaLang/julia/ui/repl.c:139
main at /home/travis/build/JuliaLang/julia/ui/repl.c:264
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4015f0)
92.963630make[2]: *** [/home/travis/build/JuliaLang/julia/usr/lib/julia/sys.o] Error 1 *** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
```

In AppVeyor the error looks like this:
```
error during bootstrap:
LoadError(at "sysimg.jl" line 315: LoadError(at "REPL.jl" line 415: Base.MethodError(f=Base.Terminals.jl
#isless(LineEdit.jl
REPLCompletions.jl
), REPL.jl
args=(-1, nothing), world=0x00002f0f)))
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [Makefile:228: /c/projects/julia/usr/lib/julia/sys.o] Error 1
make[1]: Leaving directory '/c/projects/julia'
make: *** [Makefile:105: julia-sysimg-release] Error 2
```

alyst commented:
So it fails, at `reduce.jl:425` (`while i <= last`) when called as
```
Base.mapreduce_impl(Base.operator_precedence, Base.scalarmin, Any[:(<=), :(<=)], 1, 2)
```
during bootstrap.
In the latest version of the PR this line is `@simd for i in i:last`, but there is still a similar error (see the logs).
I guess removing the `n < 16` case from `_mapreduce()` exposed what looks to me a compilation/lowering issue in `mapreduce_impl()` for max/min.
@tkelman could you please have a look?

alyst committed:
reduce code duplication in _mapreduce()

- let mapreduce_impl() handle all but empty collection cases
- don't test mapreduce_impl() directly
- extend mapreduce() tests with empty/long collections

alyst commented:
Ok, it looks like the last error was because `UnitRange` was not supported during bootstrap.
It's a pity, because it can allow `@simd for`, which may boost performance.
Of course, `mapreduce_impl()` can be defined as post-bootstrap only, so  `@simd for` would be allowed there and not in `_mapreduce()`, and `_mapreduce()` would support small array without calling `mapreduce_impl()`.
However, the behaviour of `mapreduce_impl(f, min/max)` and the internal `_mapreduce()` implementation were different (skipping vs not skipping initial NaNs), meaning that for certain operators `mapreduce()` may have different behaviour depending on the array length.

alyst commented:
So now it fails in tests that expect `maximum(NaN-containing array) = NaN`.
I guess that means the [current] behaviour of `mapreduce_impl(f, min/max, ...)` have to be fixed.

alyst committed:
mapreduce_impl(f, min/max): don't skip NaNs

- update mapreduce_impl(f, min/max) to 0.6 behaviour of treating NaNs,
  also stop scanning the array once NaN is detected
- add NaN minimum/maximum() tests for 2-element and long arrays

alyst commented:
@tkelman As now the PR makes maximum/minimum work consistently both for small and large arrays, maybe it's worth merging it to evaluate the impact on the packages?

tkelman commented on base/reduce.jl:
```diff
@@ -415,17 +399,11 @@ function mapreduce_impl(f, op::Union{typeof(scalarmax),
                         A::AbstractArray, first::Int, last::Int)
     # locate the first non NaN number
-    v = f(A[first])
+    @inbounds v = f(A[first])
```
it's not safe to assume any arbitrary function can be run without bounds checks

alyst committed:
mapreduce(f, op): don't apply @inbounds for `f`

alyst commented:
@tkelman do you think it's ready?

tkelman commented on base/reduce.jl:
```diff
@@ -168,14 +168,17 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
 function mapreduce_impl(f, op, A::AbstractArray, ifirst::Integer, ilast::Integer, blksize::Int=pairwise_blocksize(f, op))
-    if ifirst + blksize > ilast
+    if ifirst == ilast
+        @inbounds a1 = A[ifirst]
```
does every caller of mapreduce_impl guarantee that ifirst through ilast are valid indices?

alyst replied:
@tkelman I've added a note about `mapreduce_impl()` usage.

P.S. I appreciate your rigorous review as it definitely improves both the PR and my understanding of Julia. ;)

alyst commented:
ATM the only caller is `_mapreduce()` with `linearindices()`, so the range should always be valid. But to warn any future/external users we may add a note to `mapreduce_impl()` stating that it is a caller responsibility to check the range.

alyst committed:
clarify how mapreduce_impl() should be used

alyst commented:
@tkelman Thanks! Should I prepare a backport PR for 0.5 (excluding the minimum/maximum behaviour change for NaNs)?

KristofferC commented:
Thanks for sticking out with this!

tkelman commented:
Knew I forgot something, should've run nanosoldier here, looks like some big regressions in the daily run.

alyst commented:
Oops, :(
I've checked type stability of a few `_mapreduce()`/`mapreduce_impl()` calls and it looks ok.
Could it be because those test operate with small (<=16) arrays, which were previously handled entirely within `_mapreduce()` and now it is handled by `mapreduce_impl()`, so the benchmarks are bad because they include `mapreduce_impl()` compilation times? Otherwise it's hard to explain infinite memory growth.

alyst commented:
I can return the small array case to see if the benchmarks would be back to normal.

tkelman commented:
Worth trying.
